pkg/cnb/mountable_buildpack_layer_test.go

```diff
@@ -0,0 +1,165 @@
+
+	"github.com/pivotal/kpack/pkg/apis/experimental/v1alpha1"
+)
+
+func TestMountableBuildpackLayer(t *testing.T) {
+	spec.Run(t, "Test Mountable Buildpack Layer", testMountableBuildpackLayer)
+}
+
+func testMountableBuildpackLayer(t *testing.T, when spec.G, it spec.S) {
+	handler := http.NewServeMux()
+	server := httptest.NewServer(handler)
+	tagName := fmt.Sprintf("%s/some/image:tag", server.URL[7:])
+
+	var registryCalls []string
+	var backingRandomLayer v1.Layer
+	var layer v1.Layer
+
+	it.Before(func() {
+		var err error
+
+		backingRandomLayer, err = random.Layer(1000, types.DockerLayer)
+		require.NoError(t, err)
+
+		handler.HandleFunc("/v2/", func(writer http.ResponseWriter, request *http.Request) {
+			registryCalls = append(registryCalls, request.RequestURI)
+
+			writer.WriteHeader(200)
+		})
+
+		digest, err := backingRandomLayer.Digest()
+		require.NoError(t, err)
+
+		handler.HandleFunc(fmt.Sprintf("/v2/some/image/blobs/%s", digest.String()), func(writer http.ResponseWriter, request *http.Request) {
+			registryCalls = append(registryCalls, request.RequestURI)
+
+			uncompressed, err := backingRandomLayer.Compressed()
+			require.NoError(t, err)
+
+			io.Copy(writer, uncompressed)
+		})
+
+		diffID, err := backingRandomLayer.DiffID()
+		require.NoError(t, err)
+
+		size, err := backingRandomLayer.Size()
+		require.NoError(t, err)
+
+		layer, err = layerFromStoreBuildpack(authn.DefaultKeychain, v1alpha1.StoreBuildpack{
+			BuildpackInfo: v1alpha1.BuildpackInfo{
+				ID:      "id.buildpack",
+				Version: "1.0",
+			},
+			DiffId: diffID.String(),
+			StoreImage: v1alpha1.StoreImage{
+				Image: tagName,
+			},
+			Digest: digest.String(),
+			Size:   size,
+		})
+		require.NoError(t, err)
+	})
+
+	it("is a MountableLayer", func() {
+		assert.IsType(t, &remote.MountableLayer{}, layer)
+
+		expectedReference, err := name.ParseReference(tagName)
+		require.NoError(t, err)
+
+		assert.Equal(t, expectedReference, layer.(*remote.MountableLayer).Reference)
+	})
+
+	it("Digest()", func() {
+		digest, err := layer.Digest()
+		require.NoError(t, err)
+
+		expectedDigest, err := backingRandomLayer.Digest()
+		require.NoError(t, err)
+
+		assert.Equal(t, expectedDigest, digest)
+		assert.Empty(t, registryCalls)
+	})
+
+	it("Size()", func() {
+		size, err := layer.Size()
+		require.NoError(t, err)
+
+		expectedSize, err := backingRandomLayer.Size()
+		require.NoError(t, err)
+
+		assert.Equal(t, expectedSize, size)
+		require.Empty(t, registryCalls)
+	})
+
+	it("DiffID()", func() {
+		diffID, err := layer.DiffID()
+		require.NoError(t, err)
+
+		expectedDiffId, err := backingRandomLayer.DiffID()
+		require.NoError(t, err)
+
+		assert.Equal(t, expectedDiffId, diffID)
+		require.Empty(t, registryCalls)
+	})
+
+	it("MediaType()", func() {
+		mediaType, err := layer.MediaType()
+		require.NoError(t, err)
+
+		assert.Equal(t, types.DockerLayer, mediaType)
+		require.Empty(t, registryCalls)
+	})
+
+	it("Compressed()", func() {
```

Comment:
Can we add a test that makes the Compressed() and/or Uncompressed() calls twice or a call to Compressed() and then Uncompressed() to ensure that it makes the call to remote just once?

Reply:
The backing remote.Layer will actually call the registry multiple times on each call of `Compressed` or `Uncompressed`. (Doesn't store the layer contents in memory) 

The memoization only prevents multiple calls to `/v1/` (for auth) again.  

So, I am inclined to not bother testing it.